torch/csrc/jit/codegen/cuda/kernel.h

```diff
@@ -13,44 +14,94 @@ namespace torch {
 
-// Container for a lowered Kernel IR
+//! Summary of interesting facts about the kernel
 //
-// TODO(kir): currently, it is just pointing to nodes owned
-//  by a Fusion object. The goal is to have the Kernel object
-//  own the Kernel IR nodes
+// TODO(kir): const node ptrs
 //
+struct KernelSummary {
+  //! List of global buffers
+  std::vector<kir::Allocate*> global_allocations;
+
+  //! List of dynamic shared memory buffers
+  std::vector<kir::Allocate*> dynamic_smem_allocations;
+
+  //! List of static shared memory buffers
+  std::vector<kir::Allocate*> static_smem_allocations;
+
+  //! Indicate the need to generate random numbers
+  bool is_stochastic = false;
+
+  //! Do we have any block reductions?
+  bool has_block_reductions = false;
+
+  //! Do we have any grid reductions?
+  bool has_grid_reductions = false;
+
+  //! Do we have any block broadcasts?
+  bool has_block_broadcasts = false;
+
+  //! Largest shared memory buffer base type
+  DataType largest_smem_data_type = DataType::Null;
+};
+
+//! Container for a lowered Kernel IR
+//!
+//! TODO(kir): currently, it is just pointing to nodes owned
+//!  by a Fusion object. The goal is to have the Kernel object
+//!  own the Kernel IR nodes
+//!
 class TORCH_CUDA_API Kernel final : public NonCopyable {
  public:
-  explicit Kernel(const std::vector<Expr*>& exprs);
+  Kernel(std::vector<Expr*> exprs, ThreadPredicateMap predicate_map);
 
-  const auto& globalAllocations() const {
-    return global_allocations_;
+  // Register input as an input of the kernel
+  void addInput(Val* input) {
+    inputs_.push_back(input);
   }
 
-  const auto& dynamicAllocations() const {
-    return dynamic_smem_allocations_;
+  // Register output as an output of the kernel
+  void addOutput(Val* output) {
+    outputs_.push_back(output);
   }
 
-  const auto& staticAllocations() const {
-    return static_smem_allocations_;
+  const auto& inputs() const {
+    return inputs_;
+  }
+
+  const auto& outputs() const {
+    return outputs_;
   }
 
   const auto& exprs() const {
     return exprs_;
   }
 
- private:
-  // List of global buffers
-  std::vector<kir::Allocate*> global_allocations_;
+  const KernelSummary& summary() const {
+    return summary_;
+  }
 
-  // List of dynamic shared memory buffers
-  std::vector<kir::Allocate*> dynamic_smem_allocations_;
+  const ThreadPredicateMap& predicateMap() const {
+    return predicate_map_;
+  }
 
-  // List of static shared memory buffers
-  std::vector<kir::Allocate*> static_smem_allocations_;
+ private:
+  // Analyze the kernel IR and caches the summary of interesting data
+  void analyze();
 
+ private:
   // Lowered expressions
   std::vector<Expr*> exprs_;
+
+  // Kernel inputs and outputs
+  std::vector<Val*> inputs_;
+  std::vector<Val*> outputs_;
+
+  // Summary of interesting kernel data
+  KernelSummary summary_;
+
+  // Predicate map
+  // TODO(kir): consider a simpler, kernel IR based version
+  ThreadPredicateMap predicate_map_;
```

Comment:
IIRC, the only reason we need to keep this mapping is for code generation of broadcastOp. A device function, `blockBroacast`, must be used when broadcasting thread-parallelized dimensions. Whether we should call that function is currently only determined at the code-gen time, but really I think this should be captured when lowering to KIR. One idea may be to have `BlockBroadcast` KIR node and generate that KIR node instance when FIR is lowered to KIR.

Reply:
>  One idea may be to have BlockBroadcast KIR node and generate that KIR node instance when FIR is lowered to KIR.

I really like this idea. In general, I think this is the right pattern for conditional code generation: generate the intended operations during lowering rather than deciding what to print at the last minute.